src/intersection-observer.js

```diff
@@ -124,9 +125,17 @@ export class IntersectionObserver extends Observable {
     // The second time this is called, it doesn't do much but it
     // guarantees that the receiver gets an initial intersection change
     // record.
-    listenFor(this.iframe_, 'send-intersections', () => {
+    listenFor(this.iframe_, 'send-intersections', (data, source, origin) => {
+      // This message might be from any window within the iframe, we need
+      // to keep track of which windows want to be sent updates.
+      if (!this.clientWindows_.some(entry => entry.win == source)) {
```

Comment:
This is in fact the correct place. The point of these changes is to allow nested frames within a given 3P frame to have access to intersection observer updates even when they do not have access to the existing context API.

Interested windows within the 3P frame send their send-intersections messages to register interest. The code in iframe-helper.js ensures that these windows are actually within that frame and that they know the correct per-frame sentinel. It is then up to IntersectionObserver to keep track of these interested windows and broadcast updates to them.